### PR TITLE
Update on SP Factorization

### PR DESCRIPTION
**Pull request by HaiwangYu**
This is a work in progress. Making this PR for easier collaborative development.
Corresponding configuration:
https://github.com/HaiwangYu/wct-analysis/blob/master/fac-sigproc/

Summary
- `TensorSetUnpacker`, `TensorPacker` and their `iface` base classes
- updated `Decon2DResponse` and `Decon2DFilter`
- channel mask map saved in `TaggedFrameTensorSet`
- storage order changed in converters in `aux/Util`

Related talks:
- https://indico.bnl.gov/event/8214/contributions/37195/
- https://indico.bnl.gov/event/8215/contributions/37413/
- https://indico.bnl.gov/event/8216/contributions/37608/

### CODE REVIEW

File: aux/src/TaggedFrameTensorSet.cxx [hunk 1]
```diff
@@ -154,12 +155,50 @@ bool Aux::TaggedFrameTensorSet::operator()(const input_pointer& in, output_point
...
+                auto bl = ch_bl.second;
+                for(auto b : bl) {
+                    ranges_arr(ind, 0) = b.first;
+                    ranges_arr(ind, 1) = b.second;
+                    channels_arr(ind) = ch;
+                    ++ind;
+                }
+            }
+
+            auto& ranges_md = ranges->metadata();
```
Comment (brettviren) [inline aux/src/TaggedFrameTensorSet.cxx:184]: I think this metadata schema is okay but could be better "normalized".  Anytime I find myself adding an "_" to a key or constructing a value with delimiters I know I'm creating something that isn't normalized.

The issue is that the receiving code will have to be somewhat intelligent to parse and interpret the `type` value as constructed here.  If we use well normalized metadata then all of the parsing is handled and the interpretation is minimized.

We may use deep structures here so I suggest a root on a single top-level key `cmm` to hold a structure with key `type` (value "range" or "channel") and key `label` with the label.  Ie, 

```c++
ranges_md["cmm"]["type"] = "range";
ranges_md["cmm"]["label"] = label;
```

(Eventually this data may be converted to HDF5 and we need an algorithm that flattens this structure as HFD5 attributes must be flat.  I may have added support for this already, if not, it's an easy fix.)

File: sig/src/Decon2DFilter.cxx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+    auto ch_ten = Aux::get_tens(in, tag, "channels");
+    if (!ch_ten) {
+        log->error("Tensor->Frame: failed to get channels tensor for tag {}", tag);
+        THROW(ValueError() << errmsg{"Sig::Decon2DFilter::operator() !ch_ten"});
+    }
+    Eigen::Map<const Eigen::ArrayXi> ch_arr((const int *) ch_ten->data(), nchans);
+
+    bool have_cmm = true;
+    auto cmm_range = Aux::get_tens(in, tag, "bad:cmm_range");
```
Comment (brettviren) [inline sig/src/Decon2DFilter.cxx:67]: The "bad" part of this string should be configurable with a default value of "bad".  This need not be done immediately.

File: sig/src/Decon2DFilter.cxx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    auto cmm_channel = Aux::get_tens(in, tag, "bad:cmm_channel");
+    if (!cmm_range or !cmm_channel) {
+        log->debug("Tensor->Frame: no optional cmm_range tensor for tag {}", tag);
+        have_cmm = false;
+    }
+
+    int iplane = get<int>(m_cfg, "iplane", 0);
+    log->debug("iplane {}", iplane);
+
+    const std::vector<std::string> filter_names{"Wire_ind", "Wire_ind", "Wire_col"};
```
Comment (brettviren) [inline sig/src/Decon2DFilter.cxx:77]: I don't see this used in the patch.  My concern is what these strings mean and where they come from.  They may also need to be provided as configuration (eventually).

**Comment by HaiwangYu** [comment]
Thanks @brettviren for the suggestions.
I will streamline the `metadata` scheme and configuration of the nodes in the next update.

**Pull request merged**